### ISSUE TITLE
An option to enforce certain line endings (LF / CRLF)

### ISSUE THREAD

**Issue by kachkaev** [prettier/prettier#5320]
My issue is opposite to #92, which was closed by #472.

Our team uses Macs and Windows machines to work on the same code base and we struggle to enforce Unix-style line endings in the central repo. Time and again a person on Windows commits a file with CRLF instead of LF despite that we're using multiple methods to prevent that. These include:

* `git config --global core.autocrlf false` in contributing instructions (this does not help if the repo has already been checked-out with CRLF)
* `* text=auto eol=lf` in the repo's `.gitattributes` (not sure how this works)
* `.vscode/settings.json` that includes `"files.eol": "\n"` (not everyone uses vscode)
* `husky` + `lint-staged` as in [Prettier instructions](https://prettier.io/docs/en/precommit.html) (this helps commit clean code, but does not fix line endings)
* A CI/CD task that runs [`prettier-check`](https://www.npmjs.com/package/prettier-check) for the same file pattern as we have in `lint-staged` (it passes even if some files use lf and some crlf)

All the above measures do not stop a contributor from creating a new file with CRLF and committing it as is. Not setting `git config --global core.autocrlf false` moves the issue to another place: now a Windows user can check out the files in cmd.exe or SourceTree (LF → CRLF on checkout) and commit them via Ubuntu terminal (CRLF are preserved).

Prettier could really help us here if it had an option to enforce a certain flavour of line endings. I know that new options are not welcomed in general, but I can't really see a good workaround. The only thing that comes to my mind is to tweak our `package.json` somewhat like this:

```diff
   "lint-staged": {
     "complex-file-pattern-for-files-that-get-prettified": [
       "prettier --write",
+      "some-npm-module-that-converts-crlf-to-lf-for-a-given-file",
       "git add"
     ]
   },
```

This will serve as an additional barrier to not committing CRLF, however, it's not 100% robust. Someone can disable git hooks once and thus get CRLFs trough. `prettier-check` task in CI will not catch that. CRLFs will live in the central repo until someone check outs the latest commit, searches for `\r` in all code and makes another commit.

It feels like setting the flavour of new lines is a part of making the code prettier. So what do you think folks?

```bash
prettier --eol=lf path/to/file.js
```
```yaml
# .prettierrc or .prettierrc.yaml
eol: lf
```

Accepted `eol` values: `lf`, `crlf`, `auto` (default).

**PS**: If there is any chance to get this approved and squeezed into upcoming 1.15, I'll do my best to help!

**Comment by kachkaev** [comment]
Not having CRLF in the codebase is not just a matter of style and consistency. Imagine a JS file that starts with this:

```
#!/usr/bin/env node
```

If someone accidentally adds `\r` to this first line and Prettier keeps it, the OS will attempt to launch `node\r` instead of `node` and fail.

**Comment by kachkaev** [comment]
FYI: Here's what worked for me as `some-npm-module-that-converts-crlf-to-lf-for-a-given-file`

```bash
yarn add --dev repalce
```

```json
   "lint-staged": {
     "complex-file-pattern-for-files-that-get-prettified": [
       "replace '\\r\\n' '\n'",
       "prettier --write",
       "git add"
     ]
   },
```

It reduces the probability of CRLFs in the central repo, but does not completely eliminate it (hooks can be disabled).

**Comment by duailibe** [comment]
@kachkaev I get the sentiment, but I personally don't think it's something that Prettier should do? Basically the argument is "it would be easier for users to make Prettier do this" which is not enough of a reason. This seems very generic, and would be perfect for another tool to do. I think one could write that very very easily.

As for CI/CD, I think you just need an additional `check-crlf` task (or make the `prettier-check` become `various-checks` that uses prettier and `some-npm-module-that-converts-crlf-to-lf-for-a-given-file`.

I'm closing this, but if another maintainer disagree, let's reopen and discuss.

**Comment by kachkaev** [comment]
Hey @duailibe 👋 , thanks for sharing your thoughts!

I'd be totally with you about using a separate tool if there was some robust industry standard for this problem (e.g. on the git level). However, it seems that such solution does not exist and bringing it in as a new tool that's running before or after Prettier would be also tricky.

Imagine that for this very common problem I come up with a bespoke npm module called `crlf-fixer`. Here's how I would need to inject it into the project’s `package.json`:

```diff
 {
   "scripts": {
     "format": "prettier --write 'A_COMPLEX_GLOB_MATCHING_SRC_AND_DOCS'",
+    "format": "prettier --write 'A_COMPLEX_GLOB_MATCHING_SRC_AND_DOCS' && crlf-fixer --write 'A_COMPLEX_GLOB_MATCHING_SRC_AND_DOCS'",
     "testFormat": "prettier-check --write 'A_COMPLEX_GLOB_MATCHING_SRC_AND_DOCS'",
+    "testFormat": "prettier-check --write 'A_COMPLEX_GLOB_MATCHING_SRC_AND_DOCS' && crlf-fixer --verify 'A_COMPLEX_GLOB_MATCHING_SRC_AND_DOCS'",
   },
   "lint-staged": {
     "A_COMPLEX_GLOB_MATCHING_SRC_AND_DOCS": [
       "prettier --write",
+      "crlf-fixer --write",
       "git add"
     ]
   },
   "devDependencies": {
     "prettier": "^1.14.3",
+    "crlf-fixer": "^42.42.42",
     "...": "..."
   }
 }
```

That's quite a lot of repetition and additional project setup steps, isn't it? Would not it be nice to just change `.prettierrc.js`?

```diff
 module.exports = {
   trailingComma: "all",
+  eol: "lf",
 }
```

Interestingly, this `eol: "lf",` is exactly what prettier was doing before https://github.com/prettier/prettier/issues/92 (this proposal to keep CRLFs got rather controversial reactions, but still went through). I do understand the benefits of resolving that issue for Windows users who have `git config --global core.autocrlf true`, but it does come at a cost of squeezing CRLFs into the central repo and not noticing this until something breaks (e.g. `#!/usr/bin/env node\r`)

Just out of curiosity, I cloned a couple of public projects that use Prettier and noticed that even in those experienced communities people still end up with CRLFs in the committed code (likely contributed by Windows users who have issues with local git config).

https://github.com/apollographql/apollo-client
![screenshot 2018-11-01 at 16 51 10](https://user-images.githubusercontent.com/608862/47873985-7e9c1e80-de0a-11e8-9edd-10d81a2f535a.png)

https://github.com/apollographql/redux
![screenshot 2018-11-01 at 16 52 56](https://user-images.githubusercontent.com/608862/47874021-9a072980-de0a-11e8-9416-57e8dc00d022.png)

(note that the files with `\r` are covered by their linting scripts or hooks)

Even Prettier itself has got quite a few files with CRLF, and far not all of them have this on purpose to test Windows behaviour:
![screenshot 2018-11-01 at 16 39 05](https://user-images.githubusercontent.com/608862/47874150-f1a59500-de0a-11e8-9a53-4dee7618dbe5.png)

Without https://github.com/prettier/prettier/pull/472 merged, this hidden inconsistency in the formatting of the shared code would not be the case.

Prettier's current choice for line endings resides on the presence of CRLF/LF after the first line in a particular file. This would be fine if git's CRLF to LF mapping always worked perfectly on commits, but unfortunately this is not the case. Given that Prettier does change line endings today already (by bringing all newlines to a single format in a given file), I still feel that it's reasonable to introduce an `eol` option to this tool. WDYT?

**Comment by lydell** [comment]
Wouldn't you want to clrf-fix more files than Prettier can format?

**Comment by kachkaev** [comment]
That's a fair point @lydell. Indeed, for some projects I may want to apply crlf-fix for a wider range of text files, not just those supported by Prettier. However, in practice, any project that I recently touched had all file extensions covered by Prettier 🙂 💯

**Comment by kachkaev** [comment]
After removing the default parser in https://github.com/prettier/prettier/pull/4528, Prettier simply skips all files that it cannot format. In theory, it could be an interesting idea to apply some default parser, which would replace replace CRLFs if the file is not binary and if `eol` option is not `auto` (default). That's just a side-thought.

**Comment by lydell** [comment]
Not .gitignore.

**Comment by lydell** [comment]
Isn't this what you're looking for?

```
git ls-files -z | xargs --null crlf-fixer --verify
```

**Comment by evilebottnawi** [comment]
@kachkaev Just idea, maybe you can implement this tool as plugin for `prettier`

**Comment by evilebottnawi** [comment]
I can help with this :smile:

**Comment by evilebottnawi** [comment]
@kachkaev 
> When I was writing prettier-plugin-elm it was only possible to define new grammars and not hack into existing flows. Has this changed? Can a plugin inject into every file passed to prettier as a glob? If so, I’d be keen to draft smth!

Need check, we can implement this if that doesn't work, it's not hard :smile: 

> A plugin would work for me equally well as an option! Shall it be called prettier-plugin-eol? :)

Hm, we have a problem with `BOM` and i want to solve this too :smile: . Maybe we can implement other plugin for this, but i think we can union this in one plugin. Maybe `prettier-plugin-bytecode` or `prettier-plugin-character`? I am bad in naming :smile:

**Comment by kachkaev** [comment]
@evilebottnawi this kind of plugin that can affect any file passed to Prettier reminds me of https://github.com/prettier/prettier/issues/4424.

In an ideal world, my Prettier glob would cover all project contents rather than certain extensions in specific subfolders. The resulting file list would include stuff like `.gitignore`, but obviously exclude things I declared in `.prettierignore`. So calling `prettier --write '**/*.*'` would clean-up all ASTs when possible or simply ensure line endings / BOM (even for files with known extensions, but syntax errors).

I'm not sure about joining CRLF and BOM fixing into one plugin – these concerns seem quite distant. Could those be two plugins? Or maybe it's something that's needed often enough to become a part of Prettier core? 😅

**Comment by kachkaev** [comment]
Speaking of new lines, the funny thing is that it's already an option in Prettier core:
https://github.com/prettier/prettier/blob/master/src/main/core.js#L87

The value is just always being automatically derived by `guessLineEnding(text)`:
https://github.com/prettier/prettier/blob/master/src/main/core.js#L21-L27

This means that the implementation of `eol: auto | crlf | lf | cr?` would be a pretty easy task. The solution would not cover unknown files like `.gitignore`, but perhaps there's scope for a _null-plugin_ that would help prettier swallow those 🤔

**Comment by evilebottnawi** [comment]
>  Or maybe it's something that's needed often enough to become a part of Prettier core? sweat_smile

Not sure, looks like a war `LF vs CRLF` and someone who want `CR`. On the other hand it is prettier and just want to run and forgot about spaces, newlines, and other stuff. `LF`/`CRLF`/`CR` looks like this stuff. 

There are some guys in our team works on windows and sometimes it is pain, sometimes `CRLF` falls into our code base (in some cases, I don’t even understand how this happens as we have linters everywhere and checks, seems somebody love `push` with `--force` and skip hooks). 
Probably this is the case when a new option is justified.

/cc @vjeux 

> The solution would not cover unknown files like .gitignore, but perhaps there's scope for a null-plugin that would help prettier swallow those thinking

Solutions:
- null-plugin
- use `**/*` for prettier (he skips unknown extension, only fix `CRLF`/`CR`/`LF`), by the way it increases zero configuration, just add command `prettier .` in `scripts` and all looks pretty. Anyway it is sound as breaking chage for CLI and we can do this only in `2.0`.

I think it is bit difference issue and offer to touch her later.

**Comment by duailibe** [comment]
> That's quite a lot of repetition and additional project setup steps, isn't it? Would not it be nice to just change .prettierrc.js?

I don't want to sound demeaning but we can't expect Prettier to cover all use cases and valid requirements people have. We're a code formatting, I think this doesn't fall inside Prettier's scope.

As for the repetition and additional project setup steps, just consolidate both prettier + other tools in a single script. I mean, that's why people created tools like `prettier-eslint`. I guess you can write `prettier-crlf`?

**Comment by evilebottnawi** [comment]
@duailibe doesn't sound like zero configuration, i don't want install something new just to change `CRLF` to `LF` (it is a very simple task). We use 80 print width for better reading and `diff`, but when you have `LF` and somebody change to `CRLF` accidentally, `diff` looks very bad.

**Comment by kachkaev** [comment]
With a new `eol` option things will work just as before for everyone; we can assume that it's `eol: auto` configured for everyone at the moment. The new config option will just make it possible to enforce `eol: lf` as an opt-in for those who are concerned about the problem.

I'm currently experiencing the same pain as @evilebottnawi mentions:

> There are some guys in our team works on windows and sometimes it is pain, sometimes `CRLF` falls into our code base (in some cases, I don’t even understand how this happens as we have linters everywhere and checks, seems somebody love `push` with `--force` and skip hooks).
Probably this is the case when a new option is justified.

Those public repos [I shared above](https://github.com/prettier/prettier/issues/5320#issuecomment-435159024) suggest that this issue is not so rare.

> i just want to run and forgot about spaces, newlines, and other stuff. LF/CRLF/CR looks like this stuff.

💯 agree

**Comment by duailibe** [comment]
Whatever, this is not a hill I'm willing to die on; it just seemed pretty obvious to me that this is orthogonal to Prettier itself, but maybe I'm wrong.

FTR ESLint can fix this https://eslint.org/docs/rules/linebreak-style

**Comment by kachkaev** [comment]
@duailibe thanks for reopening this 😉 We're using ESlint's `linebreak-style` in our code as one CRLF precautions, but this only covers JS files. Prettier's scope is way bigger than that, especially given its growing plugin ecosystem 🥇

(this is unrelated, but I'm really exited about the upcoming mdx and html support 🙌 )

**Comment by evilebottnawi** [comment]
@kackaev for better zero configuration we should respect editorconfig line ending option

**Comment by kachkaev** [comment]
@evilebottnawi I did not know that Prettier reads stuff from [`.editorconfig`](https://editorconfig.org/) – that's interesting. If so, I agree that the value for `eol` can be grabbed from it (it's called `end_of_line` there).

We are not using `.editorconfig` file in our code base as it requires every contributor to have a plugin in their editor. Requiring it is quite challenging given the variety of tools and different backgrounds team members have (contributions can even come from a web UI by just copy-pasting something). In any case, `.editorconfig` alone does not help us enforce things via git hooks or detect CRLFs in CI/CD, but it does not mean that others cannot take `eol` value from it.

I can submit a proof-of-concept PR for the EOL option if there are no strong objections (`eol: auto | lf | cr | crlf`). No worries if it gets rejected later on! If we come up with any other solution that would help repo owners avoid unwanted CRLFs in git history, I'll be totally happy to take a clean and robust alternative on board 😉

**Comment by evilebottnawi** [comment]
@kachkaev feel free to send a PR with PoC

**Comment by kachkaev** [comment]
➡️ https://github.com/prettier/prettier/pull/5327

**Comment by kachkaev** [comment]
All tests pass in https://github.com/prettier/prettier/pull/5327. Changing PR status from WIP to RFC 🎉

**Comment by ikatyang** [comment]
Fixed in #5327.
